lib/honeybadger/events_worker.ex

```diff
@@ -0,0 +1,219 @@
+  def terminate(_reason, %State{} = state) do
+    Logger.debug("[Honeybadger] Terminating with #{total_event_count(state)} events unsent")
+    _ = flush(state)
+    :ok
+  end
+
+  @spec flush(State.t()) :: {:noreply, State.t(), pos_integer()}
+  defp flush(state) do
+    cond do
+      state.queue == [] and :queue.is_empty(state.batches) ->
+        {:noreply, state, current_timeout(state)}
+
+      state.queue == [] ->
+        attempt_send(state)
+
+      true ->
+        batches = :queue.in(%{batch: Enum.reverse(state.queue), attempts: 0}, state.batches)
+        attempt_send(%{state | queue: [], batches: batches})
+    end
+  end
+
+  @spec attempt_send(State.t()) :: State.t()
+  # Sends pending batches, handling retries and throttling
+  defp attempt_send(%State{} = state) do
+    {new_batches_list, throttling} =
+      Enum.reduce(:queue.to_list(state.batches), {[], false}, fn
+        # If already throttled, skip sending and retain the batch.
+        b, {acc, true} ->
+          {acc ++ [b], true}
+
+        %{batch: batch, attempts: attempts} = b, {acc, false} ->
+          case state.send_events_fn.(batch) do
+            :ok ->
+              Logger.debug("[Honeybadger] Sent batch of #{length(batch)} events.")
+              {acc, false}
+
+            {:error, reason} ->
+              throttling = reason == :throttled
+              updated_attempts = attempts + 1
+
+              if throttling do
+                Logger.warning(
+                  "[Honeybadger] Rate limited (429) events - (batch attempt #{updated_attempts}) - waiting for #{state.throttle_wait}ms"
+                )
+              else
+                Logger.debug(
+                  "[Honeybadger] Failed to send events batch (attempt #{updated_attempts}): #{inspect(reason)}"
+                )
+              end
+
+              if updated_attempts < state.max_batch_retries do
+                {acc ++ [%{b | attempts: updated_attempts}], throttling}
+              else
+                Logger.debug(
+                  "[Honeybadger] Dropping events batch after #{updated_attempts} attempts."
+                )
+
+                {acc, throttling}
+              end
+          end
+      end)
+
+    current_time = System.monotonic_time(:millisecond)
+
+    # Log dropped events if present and we haven't logged within the last
+    # @dropped_log_interval
+    state =
+      if state.dropped_events > 0 and
+           current_time - state.last_dropped_log >= @dropped_log_interval do
+        Logger.info("[Honeybadger] Dropped #{state.dropped_events} events due to max queue limit")
+        %{state | dropped_events: 0, last_dropped_log: current_time}
+      else
+        state
+      end
+
+    new_state = %{state | batches: :queue.from_list(new_batches_list), throttling: throttling}
+
+    {:noreply, new_state, current_timeout(new_state)}
+  end
+
+  @spec total_event_count(State.t()) :: non_neg_integer()
+  # Counts events in both the queue and pending batches.
```

Comment:
It certainly won't hurt anything, but specs for private functions are a bit unusual.